### PR TITLE
On create kt wasm

### PR DESCRIPTION
**Pull request by SHeimlich**
Add the onCreate() method to particles in Kotlin-Wasm.

### CODE REVIEW

File: java/arcs/sdk/wasm/WasmParticleImpl.kt [hunk 1]
```diff
@@ -172,4 +172,6 @@ abstract class WasmParticleImpl {
      * @param allSynced flag indicating if all handles are synchronized
      */
     open fun onHandleSync(handle: WasmHandle, allSynced: Boolean) = Unit
+
+    open fun onCreate() = Unit
```
Comment (csilvestrini) [inline java/arcs/sdk/wasm/WasmParticleImpl.kt]: Could you add KDoc for this method?

File: javatests/arcs/sdk/wasm/OnCreateTest.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.sdk.wasm
+
+class OnCreateTest : AbstractOnCreateTest() {
+    var s = "Not created!"
+
+    override fun onCreate() {
+        s = "Created!"
```
Comment (csilvestrini) [inline javatests/arcs/sdk/wasm/OnCreateTest.kt]: I thought the point of onCreate was to write default state to handles? If that's the case, then a more complete test would be to do that here instead. So you could call `handles.fooHandle.store("Created!")` here, and remove the `onHandleSync` stuff.

File: particles/Native/Kotlin/src/wasm/arcs/platform.kt [hunk 3]
```diff
@@ -228,6 +228,7 @@ actual open class DomParticleBase<Props, State> actual constructor(
             kotlinx.serialization.json.Json.nonstrict.stringify(serializer, rawData)
         )
     }
+
```
Comment (csilvestrini) [inline particles/Native/Kotlin/src/wasm/arcs/platform.kt]: Revert?

File: src/runtime/wasm.ts [hunk 4]
```diff
@@ -872,6 +872,13 @@ export class WasmParticle extends Particle {
   // Ignored for wasm particles.
   async onHandleDesync(handle: Handle<CRDTTypeRecord>) {}
 
+  async onCreate() {
+    // TODO(heimlich): not yet implemented in CPP
```
Comment (csilvestrini) [inline src/runtime/wasm.ts]: Could you file an issue for this, and link to it here?

File: src/wasm/tests/wasm-api-test.ts [hunk 5]
```diff
@@ -524,5 +527,37 @@ Object.entries(testMap).forEach(([testLabel, testDir]) => {
...
+      const {arc, stores} = await setup('OnCreateTest');
+      const fooHandle = await singletonHandleForTest(arc, stores.get('fooHandle'));
+
+      const sendEvent = async handler => {
+        await arc.idle;
+        arc.peh.sendEvent(arc.activeRecipe.particles[0], 'root', {handler});
+        await arc.idle;
+      };
+
+      //await sendEvent('case1');
```
Comment (csilvestrini) [inline src/wasm/tests/wasm-api-test.ts]: Delete?

File: src/wasm/tests/wasm-api-test.ts [hunk 6]
```diff
@@ -524,5 +527,37 @@ Object.entries(testMap).forEach(([testLabel, testDir]) => {
         'c3:90,Curly,true',
       ]);
     });
+
+    it('onCreate() Wasm koala', async function() {
+      if (isCpp) {
```
Comment (csilvestrini) [inline src/wasm/tests/wasm-api-test.ts]: Add TODO with link to GitHub issue

File: src/wasm/tests/wasm-api-test.ts [hunk 7]
```diff
@@ -524,5 +527,37 @@ Object.entries(testMap).forEach(([testLabel, testDir]) => {
         'c3:90,Curly,true',
       ]);
     });
+
+    it('onCreate() Wasm koala', async function() {
```
Comment (csilvestrini) [inline src/wasm/tests/wasm-api-test.ts]: Test name seems... odd :P

Comment (piotrswigon) [reply src/wasm/tests/wasm-api-test.ts]: 🐨

File: javatests/arcs/sdk/wasm/OnCreateTest.kt (hunk 2 above)
Comment (piotrswigon) [reply javatests/arcs/sdk/wasm/OnCreateTest.kt]: The test that @SHeimlich is doing below is allowing her to distinguish between a boot up sequence in which onCreate() is called, vs. a sequence where onCreate() is *not* called.

If the only writing to handles happened in onCreate(), then when testing the deserialization flow and not seeing anything written, one could wonder was the particle booted at all. With the test structure like Sarah has, it is clear that the particle has booted, and that onCreate was not called (at least not before onHandleSync()).

One thing that could be improved, is ensuring that handles are ready for writing when onCreate() gets invoked. The particle could be restructured like below to achieve that:
```
class OnCreateTest : AbstractOnCreateTest() {
    var onCreateCalled = false

    override fun onCreate() {
        onCreateCalled = true;
        handles.fooHandle.store(OnCreateTest_FooHandle("Created!"))
    }

    override fun onHandleSync(handle: WasmHandle, allSynced: Boolean) {
        if (!onCreateCalled) {
          handles.fooHandle.store(OnCreateTest_FooHandle("Not created!"))
        }
    }
}
```

WDYT?

Comment (piotrswigon) [reply javatests/arcs/sdk/wasm/OnCreateTest.kt]: The remaining thing that is not explicitly tested in the deserialization flow is that onCreate() never gets called (vs. before onHandleSync), but I think we can live with it.

Comment (SHeimlich) [reply javatests/arcs/sdk/wasm/OnCreateTest.kt]: This didn't work initially due to a bug in the TS implementation. That is being resolved with #4797

File: java/arcs/sdk/wasm/WasmParticleImpl.kt (hunk 1 above)
Comment (SHeimlich) [reply java/arcs/sdk/wasm/WasmParticleImpl.kt]: Added.

File: src/wasm/tests/wasm-api-test.ts (hunk 6 above)
Comment (SHeimlich) [reply src/wasm/tests/wasm-api-test.ts]: Done

**Pull request merged**